packages/backend-storage/src/access_builder.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  AuthRoleName,
+  ConstructFactory,
+  ConstructFactoryGetInstanceProps,
+  ResourceAccessAcceptor,
+  ResourceAccessAcceptorFactory,
+  ResourceProvider,
+} from '@aws-amplify/plugin-types';
+
+export type StorageAction = 'read' | 'write' | 'delete';
+
+export type StorageAccessDefinition = {
+  getResourceAccessAcceptor: (
+    getInstanceProps: ConstructFactoryGetInstanceProps
+  ) => ResourceAccessAcceptor;
+  /**
+   * Actions to grant to this role on a specific prefix
+   */
+  actions: StorageAction[];
+  /**
+   * The value that will be substituted into the resource string in place of the {owner} token
+   */
+  ownerPlaceholderSubstitution: string;
+};
+
+export type StorageAccessBuilder = {
+  to: (...actions: StorageAction[]) => StorageAccessDefinition;
+};
+
+/**
+ * !EXPERIMENTAL!
+ *
+ * Resource access patterns are under active development and are subject to breaking changes.
+ * Do not use in production.
+ */
+export type RoleAccessBuilder = {
+  authenticated: StorageAccessBuilder;
+  guest: StorageAccessBuilder;
+  owner: StorageAccessBuilder;
+  resource: (
+    other: ConstructFactory<ResourceProvider & ResourceAccessAcceptorFactory>
+  ) => StorageAccessBuilder;
+};
+
+export const storageAccessBuilder: RoleAccessBuilder = {
+  authenticated: {
+    to: (...actions) => ({
+      getResourceAccessAcceptor: getAuthRoleResourceAccessAcceptor,
+      actions,
+      ownerPlaceholderSubstitution: '*',
+    }),
+  },
+  guest: {
+    to: (...actions) => ({
+      getResourceAccessAcceptor: getUnauthRoleResourceAccessAcceptor,
+      actions,
+      ownerPlaceholderSubstitution: '*',
+    }),
+  },
+  owner: {
+    to: (...actions) => ({
+      getResourceAccessAcceptor: getAuthRoleResourceAccessAcceptor,
+      actions,
+      ownerPlaceholderSubstitution: '${cognito-identity.amazon.com:sub}',
+    }),
+  },
+  resource: (other) => ({
+    to: (...actions) => ({
+      getResourceAccessAcceptor: (getInstanceProps) =>
+        other.getInstance(getInstanceProps).getResourceAccessAcceptor(),
+      actions,
+      ownerPlaceholderSubstitution: '*',
+    }),
+  }),
+};
+
+const getAuthRoleResourceAccessAcceptor = (
+  getInstanceProps: ConstructFactoryGetInstanceProps
+) =>
+  getUserRoleResourceAccessAcceptor(
+    getInstanceProps,
+    'authenticatedUserIamRole'
+  );
+
+const getUnauthRoleResourceAccessAcceptor = (
+  getInstanceProps: ConstructFactoryGetInstanceProps
+) =>
+  getUserRoleResourceAccessAcceptor(
+    getInstanceProps,
+    'unauthenticatedUserIamRole'
+  );
```

Comment:
Shouldn't these come from `backend-auth`? Storage shouldn't concern itself about auth's internal props. 

Auth can expose methods for storage to understand what's authenticated role and unauthenticated roles are

Reply:
those strings are coming from `AuthRoleName` exported by `plugin-types` so it's not depending on anything that isn't an exported interface. However, I have debated whether these functions should go in `platform-core` instead of here. I decided to wait until we needed them in more places

Reply:
In my opinion, they should go to the `platform-core`. Also ok waiting for this to move later since customers don't see this.

Reply:
I think I agree with you, but I'll move this later if/when we need it for other categories